File: src/formatters/pretty.ts
// Pretty terminal output with ANSI colors

import type { MessageOutput, ProcessedChannel, ProcessedMessage } from '../types'
import { bold, cyan, dim, userColor } from '../utils/colors'
import { formatRelativeTime, formatTime, getDateGroupLabel } from '../utils/date'

function channelHeader(channel: ProcessedChannel): string {
  if (channel.type === 'dm') {
    return `💬 DMs with ${cyan(channel.name)}`
  }
  if (channel.type === 'group') {
    return `💬 Group DM: ${cyan(channel.name)}`
  }
  const name = `#${channel.name}`
  const display = channel.displayName ? ` (${channel.displayName})` : ''
  return `📢 ${cyan(name)}${display}`
}

function channelHeaderPlain(channel: ProcessedChannel): string {
  if (channel.type === 'dm') {
    return `DMs with ${channel.name}`
  }
  if (channel.type === 'group') {
    return `Group DM: ${channel.name}`
  }
  const display = channel.displayName ? ` (${channel.displayName})` : ''
  return `#${channel.name}${display}`
}

export interface PrettyOptions {
  color?: boolean
  relative?: boolean
}

export function formatPretty(
  outputs: MessageOutput[],
  options: PrettyOptions | boolean = true,
): string {
  // Handle legacy boolean parameter
  const opts: PrettyOptions = typeof options === 'boolean' ? { color: options } : options
  const useColor = opts.color ?? true
  const relative = opts.relative ?? false

  if (!useColor) {
    return formatPrettyNoColor(outputs, relative)
  }

  const sections: string[] = []

  for (const output of outputs) {
    sections.push(formatChannelPretty(output, relative))
  }

  return sections.join(`\n${dim('─'.repeat(60))}\n\n`)
}

function formatChannelPretty(output: MessageOutput, relative: boolean): string {
  const { channel, messages } = output
  const lines: string[] = []

  // Header
  lines.push(bold(channelHeader(channel)))
  lines.push('')

  // Group messages by date
  const messagesByDate = groupByDate(messages)

  for (const [date, msgs] of messagesByDate) {
    lines.push(dim(`  ── ${date} ──`))
    lines.push('')

    for (const msg of msgs) {
      lines.push(formatMessagePretty(msg, relative))
    }
  }

  // Redaction notice
  if (output.redactions.length > 0) {
    lines.push('')
    lines.push(dim(`  ⚠ ${output.redactions.length} secret(s) redacted`))
  }
  appendCoverage(lines, output)

  return lines.join('\n')
}

function formatMessagePretty(
  msg: ProcessedMessage,
  relative: boolean,
  indent: string = '  ',
): string {
  const timeStr = relative ? formatRelativeTime(msg.timestamp) : formatTime(msg.timestamp)
  const time = dim(timeStr)
  const user = userColor(msg.user)
  const markers = messageMarkers(msg)

  const lines: string[] = []
  lines.push(
    `${indent}${time} ${bold(user)}${markers ? ` ${dim(markers)}` : ''} ${dim(compactPostRef(msg))}`,
  )

  // Indent message content
  const textIndent = `${indent}  `
  const indentedText = msg.text
    .split('\n')
    .map((line) => `${textIndent}${line}`)
    .join('\n')
  lines.push(indentedText)
  lines.push(dim(`${textIndent}${formatStateTimes(msg)}`))

  // File attachments
  if (msg.fileDetails.length > 0) {
    lines.push(dim(`${textIndent}📎 ${formatFiles(msg)}`))
  }
  appendRichContent(lines, msg, textIndent, (value) => dim(value))

  lines.push('')

  // Render replies
  if (msg.replies && msg.replies.length > 0) {
    for (const reply of msg.replies) {
      lines.push(formatMessagePretty(reply, relative, `${indent}  ↳ `))
    }
  }

  return lines.join('\n')
}

function formatPrettyNoColor(outputs: MessageOutput[], relative: boolean): string {
  const sections: string[] = []

  for (const output of outputs) {
    const { channel, messages } = output
    const lines: string[] = []

    lines.push(channelHeaderPlain(channel))
    lines.push('─'.repeat(40))

    const messagesByDate = groupByDate(messages)

    for (const [date, msgs] of messagesByDate) {
      lines.push(`  -- ${date} --`)
      lines.push('')

      for (const msg of msgs) {
        formatMessageNoColor(msg, relative, lines, '  ')
      }
    }

    if (output.redactions.length > 0) {
      lines.push(`  [${output.redactions.length} secret(s) redacted]`)
    }
    appendCoverage(lines, output)

    sections.push(lines.join('\n'))
  }

  return sections.join(`\n${'='.repeat(60)}\n\n`)
}

function formatMessageNoColor(
  msg: ProcessedMessage,
  relative: boolean,
  lines: string[],
  indent: string,
): void {
  const timeStr = relative ? formatRelativeTime(msg.timestamp) : formatTime(msg.timestamp)
  const markers = messageMarkers(msg)
  lines.push(
    `${indent}[${timeStr}] ${msg.user}${markers ? ` ${markers}` : ''} ${compactPostRef(msg)}`,
  )
  const textIndent = `${indent}  `
  const indentedText = msg.text
    .split('\n')
    .map((line) => `${textIndent}${line}`)
    .join('\n')
  lines.push(indentedText)
  lines.push(`${textIndent}${formatStateTimes(msg)}`)
  if (msg.fileDetails.length > 0) {
    lines.push(`${textIndent}Files: ${formatFiles(msg)}`)
  }
  appendRichContent(lines, msg, textIndent, (value) => value)
  lines.push('')

  if (msg.replies && msg.replies.length > 0) {
    for (const reply of msg.replies) {
      formatMessageNoColor(reply, relative, lines, `${indent}  > `)
    }
  }
}

function compactPostRef(msg: ProcessedMessage): string {
  const id = msg.id.length > 8 ? msg.id.slice(0, 8) : msg.id
  return `${id} ${msg.permalink}`
}

function messageMarkers(msg: ProcessedMessage): string {
  const markers: string[] = []
  if (msg.isDeleted) markers.push('[deleted]')
  else if (msg.editedAt) markers.push('[edited]')
  if (msg.isSystem) markers.push(msg.postType ? `[system:${msg.postType}]` : '[system]')
  if (msg.isPinned) markers.push('[pinned]')
  return markers.join(' ')
}

function formatFiles(msg: ProcessedMessage): string {
  return msg.fileDetails
    .map((file) => {
      const label = file.name ? `${file.name} (${file.id})` : file.id
      const details = [
        file.mime,
        file.extension,
        file.size === undefined ? undefined : `${file.size} B`,
      ]
        .filter(Boolean)
        .join(', ')
      return `${label}${details ? `, ${details}` : ''}`
    })
    .join(', ')
}

function formatStateTimes(msg: ProcessedMessage): string {
  const values = [`Updated ${msg.updatedAt.toISOString()}`]
  if (msg.editedAt) values.push(`edited ${msg.editedAt.toISOString()}`)
  if (msg.deletedAt) values.push(`deleted ${msg.deletedAt.toISOString()}`)
  return values.join('; ')
}

function appendRichContent(
  lines: string[],
  msg: ProcessedMessage,
  indent: string,
  decorate: (value: string) => string,
): void {
  for (const attachment of msg.attachments) {
    if (attachment.pretext) lines.push(decorate(`${indent}${attachment.pretext}`))
    if (attachment.title) lines.push(decorate(`${indent}Attachment: ${attachment.title}`))
    if (attachment.titleLink) lines.push(decorate(`${indent}  Link: ${attachment.titleLink}`))
    if (attachment.authorName) lines.push(decorate(`${indent}  By: ${attachment.authorName}`))
    if (attachment.text) lines.push(decorate(`${indent}  ${attachment.text}`))
    for (const field of attachment.fields ?? []) {
      lines.push(
        decorate(`${indent}  ${field.title ? `${field.title}: ` : ''}${field.value ?? ''}`),
      )
    }
    if (attachment.footer) lines.push(decorate(`${indent}  ${attachment.footer}`))
    if (attachment.fallback) lines.push(decorate(`${indent}  Fallback: ${attachment.fallback}`))
    if (attachment.color) lines.push(decorate(`${indent}  Color: ${attachment.color}`))
    if (attachment.timestamp) lines.push(decorate(`${indent}  Timestamp: ${attachment.timestamp}`))
    for (const url of [
      attachment.authorLink,
      attachment.authorIcon,
      attachment.footerIcon,
      attachment.imageUrl,
      attachment.thumbUrl,
    ]) {
      if (url) lines.push(decorate(`${indent}  ${url}`))
    }
  }
  if (msg.reactions.length > 0) {
    const reactions = msg.reactions
      .map(({ emoji, count, actors }) => {
        const names = actors.map((actor) => actor.username ?? actor.id).join(', ')
        return `:${emoji}: ${count}${names ? ` (${names})` : ''}`
      })
      .join('  ')
    lines.push(decorate(`${indent}Reactions: ${reactions}`))
  }
}

function appendCoverage(lines: string[], output: MessageOutput): void {
  const state = output.retrieval.selection.queryTruncated
  lines.push(
    `  Coverage: ${output.retrieval.selection.selectedCount} selected, ${output.retrieval.visiblePostCount} visible; query ${state === true ? 'truncated' : state === false ? 'complete' : 'completeness unknown'}`,
  )
}

function groupByDate(messages: ProcessedMessage[]): Map<string, ProcessedMessage[]> {
  const groups = new Map<string, ProcessedMessage[]>()

  const sorted = [...messages].sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime())

  for (const msg of sorted) {
    const date = getDateGroupLabel(msg.timestamp)
    if (!groups.has(date)) {
      groups.set(date, [])
    }
    groups.get(date)?.push(msg)
  }

  return groups
}

File: internal/output/pretty.go
package output

import (
	"fmt"
	"io"
	"slices"
	"strings"
	"time"
	"unicode/utf16"
)

const (
	ansiReset = "\x1b[0m"
	ansiBold  = "\x1b[1m"
	ansiDim   = "\x1b[2m"
	ansiCyan  = "\x1b[36m"
)

type PrettyOptions struct {
	Color    bool
	Relative bool
}

// MessageOutput is one independently headed channel section.
type MessageOutput struct {
	Channel    Channel
	Messages   []Message
	Redactions []Redaction
	Retrieval  Retrieval
}

// FormatPretty renders already-presented values. It deliberately performs no
// sanitization or redaction of its own.
func FormatPretty(outputs []MessageOutput, dates DateFormatter, options PrettyOptions) string {
	sections := make([]string, 0, len(outputs))
	for _, output := range outputs {
		if options.Color {
			sections = append(sections, formatColorSection(output, dates, options.Relative))
		} else {
			sections = append(sections, formatPlainSection(output, dates, options.Relative))
		}
	}
	separator := "\n" + strings.Repeat("=", 60) + "\n\n"
	if options.Color {
		separator = "\n" + dim(strings.Repeat("─", 60)) + "\n\n"
	}
	return strings.Join(sections, separator)
}

// WritePretty renders once and writes once. A short write is returned as
// io.ErrShortWrite and is never retried, since the writer's state is unknown.
func WritePretty(w io.Writer, outputs []MessageOutput, dates DateFormatter, options PrettyOptions) (int, error) {
	formatted := FormatPretty(outputs, dates, options)
	n, err := io.WriteString(w, formatted)
	if err == nil && n != len(formatted) {
		err = io.ErrShortWrite
	}
	return n, err
}

func formatColorSection(output MessageOutput, dates DateFormatter, relative bool) string {
	lines := []string{bold(colorHeader(output.Channel)), ""}
	appendMessagesByDate(&lines, output.Messages, dates, func(message Message) {
		lines = append(lines, formatColorMessage(message, dates, relative, "  "))
	}, true)
	if len(output.Redactions) > 0 {
		lines = append(lines, "", dim(fmt.Sprintf("  ⚠ %d secret(s) redacted", len(output.Redactions))))
	}
	appendCoverage(&lines, output)
	return strings.Join(lines, "\n")
}

func formatPlainSection(output MessageOutput, dates DateFormatter, relative bool) string {
	lines := []string{plainHeader(output.Channel), strings.Repeat("─", 40)}
	appendMessagesByDate(&lines, output.Messages, dates, func(message Message) {
		appendPlainMessage(&lines, message, dates, relative, "  ")
	}, false)
	if len(output.Redactions) > 0 {
		lines = append(lines, fmt.Sprintf("  [%d secret(s) redacted]", len(output.Redactions)))
	}
	appendCoverage(&lines, output)
	return strings.Join(lines, "\n")
}

func appendMessagesByDate(lines *[]string, messages []Message, dates DateFormatter, appendMessage func(Message), color bool) {
	sorted := slices.Clone(messages)
	slices.SortStableFunc(sorted, func(a, b Message) int { return a.Timestamp.Compare(b.Timestamp) })
	last := ""
	for _, message := range sorted {
		label := dates.DateGroupLabel(message.Timestamp)
		if label != last {
			if color {
				*lines = append(*lines, dim("  ── "+label+" ──"), "")
			} else {
				*lines = append(*lines, "  -- "+label+" --", "")
			}
			last = label
		}
		appendMessage(message)
	}
}

func formatColorMessage(message Message, dates DateFormatter, relative bool, indent string) string {
	timeText := dates.FormatTime(message.Timestamp)
	if relative {
		timeText = dates.FormatRelativeTime(message.Timestamp)
	}
	markers := messageMarkers(message)
	header := indent + dim(timeText) + " " + bold(userColor(message.User))
	if markers != "" {
		header += " " + dim(markers)
	}
	header += " " + dim(compactPostRef(message))
	textIndent := indent + "  "
	lines := []string{header, indentLines(message.Text, textIndent), dim(textIndent + formatStateTimes(message))}
	if len(message.FileDetails) > 0 {
		lines = append(lines, dim(textIndent+"📎 "+formatFiles(message)))
	}
	appendRichContent(&lines, message, textIndent, dim)
	lines = append(lines, "")
	for _, reply := range message.Replies {
		lines = append(lines, formatColorMessage(reply, dates, relative, indent+"  ↳ "))
	}
	return strings.Join(lines, "\n")
}

func appendPlainMessage(lines *[]string, message Message, dates DateFormatter, relative bool, indent string) {
	timeText := dates.FormatTime(message.Timestamp)
	if relative {
		timeText = dates.FormatRelativeTime(message.Timestamp)
	}
	markers := messageMarkers(message)
	header := fmt.Sprintf("%s[%s] %s", indent, timeText, message.User)
	if markers != "" {
		header += " " + markers
	}
	header += " " + compactPostRef(message)
	textIndent := indent + "  "
	*lines = append(*lines, header, indentLines(message.Text, textIndent), textIndent+formatStateTimes(message))
	if len(message.FileDetails) > 0 {
		*lines = append(*lines, textIndent+"Files: "+formatFiles(message))
	}
	appendRichContent(lines, message, textIndent, func(value string) string { return value })
	*lines = append(*lines, "")
	for _, reply := range message.Replies {
		appendPlainMessage(lines, reply, dates, relative, indent+"  > ")
	}
}

func colorHeader(channel Channel) string {
	switch channel.Type {
	case "unknown":
		return "⚠ Unknown channel (" + cyan(channel.ID) + ")"
	case "dm":
		return "💬 DMs with " + cyan(channel.Name)
	case "group":
		return "💬 Group DM: " + cyan(channel.Name)
	default:
		display := ""
		if channel.DisplayName != "" {
			display = " (" + channel.DisplayName + ")"
		}
		return "📢 " + cyan("#"+channel.Name) + display
	}
}

func plainHeader(channel Channel) string {
	switch channel.Type {
	case "unknown":
		return "Unknown channel (" + channel.ID + ")"
	case "dm":
		return "DMs with " + channel.Name
	case "group":
		return "Group DM: " + channel.Name
	default:
		display := ""
		if channel.DisplayName != "" {
			display = " (" + channel.DisplayName + ")"
		}
		return "#" + channel.Name + display
	}
}

func compactPostRef(message Message) string {
	id := message.ID
	if len(id) > 8 {
		id = id[:8]
	}
	return id + " " + message.Permalink
}

func messageMarkers(message Message) string {
	markers := make([]string, 0, 3)
	if message.IsDeleted {
		markers = append(markers, "[deleted]")
	} else if message.EditedAt != nil {
		markers = append(markers, "[edited]")
	}
	if message.IsSystem {
		if message.PostType != "" {
			markers = append(markers, "[system:"+message.PostType+"]")
		} else {
			markers = append(markers, "[system]")
		}
	}
	if message.IsPinned {
		markers = append(markers, "[pinned]")
	}
	return strings.Join(markers, " ")
}

func formatFiles(message Message) string {
	files := make([]string, 0, len(message.FileDetails))
	for _, file := range message.FileDetails {
		label := file.ID
		if file.Name != "" {
			label = file.Name + " (" + file.ID + ")"
		}
		details := make([]string, 0, 3)
		if file.MIME != "" {
			details = append(details, file.MIME)
		}
		if file.Extension != "" {
			details = append(details, file.Extension)
		}
		if file.Size != nil {
			details = append(details, fmt.Sprintf("%d B", *file.Size))
		}
		if len(details) > 0 {
			label += ", " + strings.Join(details, ", ")
		}
		files = append(files, label)
	}
	return strings.Join(files, ", ")
}

func formatStateTimes(message Message) string {
	values := []string{"Updated " + isoTime(message.UpdatedAt)}
	if message.EditedAt != nil {
		values = append(values, "edited "+isoTime(*message.EditedAt))
	}
	if message.DeletedAt != nil {
		values = append(values, "deleted "+isoTime(*message.DeletedAt))
	}
	return strings.Join(values, "; ")
}

func isoTime(value time.Time) string {
	return value.UTC().Format("2006-01-02T15:04:05.000Z")
}

func appendRichContent(lines *[]string, message Message, indent string, decorate func(string) string) {
	for _, attachment := range message.Attachments {
		if attachment.Pretext != "" {
			*lines = append(*lines, decorate(indent+attachment.Pretext))
		}
		if attachment.Title != "" {
			*lines = append(*lines, decorate(indent+"Attachment: "+attachment.Title))
		}
		if attachment.TitleLink != "" {
			*lines = append(*lines, decorate(indent+"  Link: "+attachment.TitleLink))
		}
		if attachment.AuthorName != "" {
			*lines = append(*lines, decorate(indent+"  By: "+attachment.AuthorName))
		}
		if attachment.Text != "" {
			*lines = append(*lines, decorate(indent+"  "+attachment.Text))
		}
		for _, field := range attachment.Fields {
			prefix := ""
			if field.Title != "" {
				prefix = field.Title + ": "
			}
			*lines = append(*lines, decorate(indent+"  "+prefix+field.Value))
		}
		if attachment.Footer != "" {
			*lines = append(*lines, decorate(indent+"  "+attachment.Footer))
		}
		if attachment.Fallback != "" {
			*lines = append(*lines, decorate(indent+"  Fallback: "+attachment.Fallback))
		}
		if attachment.Color != "" {
			*lines = append(*lines, decorate(indent+"  Color: "+attachment.Color))
		}
		if attachment.Timestamp != "" {
			*lines = append(*lines, decorate(indent+"  Timestamp: "+attachment.Timestamp))
		}
		for _, value := range []string{attachment.AuthorLink, attachment.AuthorIcon, attachment.FooterIcon, attachment.ImageURL, attachment.ThumbURL} {
			if value != "" {
				*lines = append(*lines, decorate(indent+"  "+value))
			}
		}
	}
	if len(message.Reactions) > 0 {
		reactions := make([]string, 0, len(message.Reactions))
		for _, reaction := range message.Reactions {
			names := make([]string, 0, len(reaction.Actors))
			for _, actor := range reaction.Actors {
				if actor.Username != "" {
					names = append(names, actor.Username)
				} else {
					names = append(names, actor.ID)
				}
			}
			value := fmt.Sprintf(":%s: %d", reaction.Emoji, reaction.Count)
			if len(names) > 0 {
				value += " (" + strings.Join(names, ", ") + ")"
			}
			reactions = append(reactions, value)
		}
		*lines = append(*lines, decorate(indent+"Reactions: "+strings.Join(reactions, "  ")))
	}
}

func appendCoverage(lines *[]string, output MessageOutput) {
	state := "completeness unknown"
	if output.Retrieval.Selection.QueryTruncated != nil {
		if *output.Retrieval.Selection.QueryTruncated {
			state = "truncated"
		} else {
			state = "complete"
		}
	}
	*lines = append(*lines, fmt.Sprintf("  Coverage: %d selected, %d visible; query %s", output.Retrieval.Selection.SelectedCount, output.Retrieval.VisiblePostCount, state))
	if output.Retrieval.Selection.NextCursor != nil && *output.Retrieval.Selection.NextCursor != "" {
		*lines = append(*lines, "  Next cursor: "+*output.Retrieval.Selection.NextCursor)
	}
}

func indentLines(value, indent string) string {
	return indent + strings.ReplaceAll(value, "\n", "\n"+indent)
}

func bold(value string) string { return ansiBold + value + ansiReset }
func dim(value string) string  { return ansiDim + value + ansiReset }
func cyan(value string) string { return ansiCyan + value + ansiReset }

func userColor(value string) string {
	colors := [...]string{ansiCyan, "\x1b[33m", "\x1b[32m", "\x1b[35m", "\x1b[34m"}
	var hash int32
	for _, r := range value {
		unit := uint16(r)
		if r > 0xFFFF {
			high, _ := utf16.EncodeRune(r)
			unit = uint16(high)
		}
		hash = hash*31 + int32(unit)
	}
	index := int64(hash)
	if index < 0 {
		index = -index
	}
	return colors[index%int64(len(colors))] + value + ansiReset
}
